### PR TITLE
Filter the source given to `Player#src` before use

### PR DESCRIPTION
**Pull request by BrandonOCasey**
## Description
* Add a util that filters all possible `Tech~SourceObjects` into a valid `Tech~SourceObject` (see tests) 
* Remove `Player#sourceList_` as it is no longer in use

## Requirements Checklist
- [x] Feature implemented / Bug fixed
- [x] If necessary, more likely in a feature request than a bug fix
  - [x] Change has been verified in an actual browser (Chome, Firefox, IE)
  - [x] Unit Tests updated or fixed
- [ ] Reviewed by Two Core Contributors

### CODE REVIEW

File: src/js/player.js [hunk 1]
```diff
@@ -2218,27 +2219,23 @@ class Player extends Component {
    *         The current video source when getting
    */
   src(source) {
-    if (source === undefined) {
```
Comment (gkatsev) [inline src/js/player.js]: why was this part removed?

Comment (gkatsev) [inline src/js/player.js]: Ah, I see. I think it's better to keep it. Why do extra work if we know for sure we can exit early here?

File: src/js/player.js [hunk 2]
```diff
@@ -2218,27 +2219,23 @@ class Player extends Component {
...
-      src = {
-        src: source
-      };
+    // filter out invalid sources and turn our source into
+    // an array of source objects
+    src = filterSource(src);
 
-      this.cache_.sources = [src];
+    // no valid sources, return the current source
+    if (!src.length) {
```
Comment (gkatsev) [inline src/js/player.js]: Should we treat `src([])` the same as `src()`?

File: src/js/utils/filter-source.js [hunk 3]
```diff
@@ -0,0 +1,49 @@
+/**
+ * @module filter-source
+ */
+import {isObject} from './obj';
+
+/**
+ * Filter out single bad source objects or multiple source objects in an
```
Comment (gkatsev) [inline src/js/utils/filter-source.js]: I don't think it's worth handling nested arrays of sources.

File: src/js/player.js [hunk 4]
```diff
@@ -2218,27 +2219,23 @@ class Player extends Component {
...
+      return this.cache_.src;
     }
 
-    this.cache_.source = src;
+    this.cache_.sources = src;
+    src = src[0];
```
Comment (gkatsev) [inline src/js/player.js]: with this, we'd only ever be able to try running through the first source. See line 2247.

File: src/js/player.js [hunk 5]
```diff
@@ -2218,27 +2219,23 @@ class Player extends Component {
...
+
+    this.changingSrc_ = true;
 
+    this.cache_.source = src;
     this.currentType_ = src.type;
```
Comment (gkatsev) [inline src/js/player.js]: just realized that we need to move/copy this down into line 2264

File: src/js/player.js (hunk 4 above)
Comment (gkatsev) [inline src/js/player.js]: Or at least, we don't have the benefit of `filterSource` for subsequent iterations and the two arrays aren't necessary in sync.

File: src/js/player.js (hunk 2 above)
Comment (BrandonOCasey) [inline src/js/player.js]: I think we want to exit early in the case where an empty source is passed, I could see logging an error, but not much else would be different.

File: src/js/utils/filter-source.js (hunk 3 above)
Comment (BrandonOCasey) [inline src/js/utils/filter-source.js]: I will remove that from the doc comments/unit tests if you like, but the code will probably stay the same either way.

File: src/js/player.js (hunk 1 above)
Comment (BrandonOCasey) [inline src/js/player.js]: yeah I guess we can keep it to do less work, I just though it would be better to have less code. I don't think that filterSource will do anything with it other than turn it into an empty array. I guess it depends on wether or not we want to log an error if an invalid source array is passed to us vs being used as a getter.

Comment (gkatsev) [inline src/js/player.js]: video.js throws the following errors:
```js
player.src([]); // No compatible source was found for this media
player.src({}); // The media could not be loaded, either because the server or network failed or because the format is not supported.
```

It looks like vjs 5 special cased empty arrays to have the custom object and with an empty object it basically tried to set an empty string and had it raise the error from the video element.

Comment (gkatsev) [inline src/js/player.js]: I think we should maintain this functionality. Though, having `player.src({})` produce the same error as `player.src([])` sounds fine too.

File: src/js/player.js (hunk 5 above)
Comment (BrandonOCasey) [inline src/js/player.js]: is there any reason not to use the `cache_` object here?

File: src/js/player.js (hunk 4 above)
Comment (BrandonOCasey) [inline src/js/player.js]: I think 2247 uses the original `source` object and not `src`

File: src/js/player.js (hunk 5 above)
Comment (gkatsev) [inline src/js/player.js]: probably no reason not to ad that in. This was just how it used to be.

File: src/js/player.js (hunk 4 above)
Comment (gkatsev) [inline src/js/player.js]: Yes, but previously `source` was the source of truth for a list of sources. With the `filterSource` functionality, the output of `filterSource` became the source of truth.

File: src/js/player.js (hunk 1 above)
Comment (BrandonOCasey) [inline src/js/player.js]: Do we want to wrap the `this.error` call in a `setTimeout` like what was done in `sourceList_`?

File: src/js/player.js (hunk 4 above)
Comment (BrandonOCasey) [inline src/js/player.js]: I see

File: src/js/player.js (hunk 5 above)
Comment (BrandonOCasey) [inline src/js/player.js]: I am wondering two things now:
1. Do we need to change `this.cache_.source`, `this.cache_.src`, and `this.cache_.type` in the `middleware.setSource` callback and also above it? or an we just do it inside?
2. Should we even cache `this.cache_.src` and `this.cache_.type` or should we just cache `this.cache_.source` and use that when we need to get the `type` or the `src` url

Comment (gkatsev) [inline src/js/player.js]: If you have a source like this:
```js
{
  src: 'foo',
  type: 'video/foo'
}
```
With a middleware that returns an m3u8 for it, we want the following behavior:
```js
player.src();
// 'http://example.com/video.m3u8'

player.currentSrc();
// 'foo'

player.currentSource();
// {src: 'foo', type: 'video/foo'}
```

Let me double check what we talked about for `currentType` previously.

Comment (gkatsev) [inline src/js/player.js]: Oh, maybe we don't need to move `currentType` because we want it to match the type of the video returned by `player.currentSrc()`. So, in the above scenario, it'll be `video/foo`.

File: src/js/player.js (hunk 1 above)
Comment (gkatsev) [inline src/js/player.js]: Probably, yes.

**Comment by BrandonOCasey** [comment]
@gkatsev this has been updated with a lot of the things that you asked about

File: src/js/player.js (hunk 1 above)
Comment (BrandonOCasey) [inline src/js/player.js]: done

File: src/js/player.js [hunk 6]
```diff
@@ -2385,14 +2357,7 @@ class Player extends Component {
    *         The current source object
    */
   currentSource() {
-    const source = {};
-    const src = this.currentSrc();
```
Comment (BrandonOCasey) [inline src/js/player.js]: This code didn't make sense because `this.currentSrc` was looking at the same variable (`this.cache_.source`).

File: src/js/player.js [hunk 7]
```diff
@@ -2218,37 +2219,39 @@ class Player extends Component {
    *         The current video source when getting
    */
   src(source) {
-    if (source === undefined) {
```
Comment (gkatsev) [inline src/js/player.js]: I really think this should be separate and at the top here. It makes a clear delineation between when it's being called as a getter vs when it's called as a setter.

File: src/js/player.js [hunk 8]
```diff
@@ -2415,7 +2380,7 @@ class Player extends Component {
    *         The source MIME type
    */
   currentType() {
-    return this.currentType_ || '';
+    return this.currentSource() && this.currentSource().type || '';
```
Comment (gkatsev) [inline src/js/player.js]: any other references to `this.currentType_` that need to be removed?

File: src/js/player.js (hunk 7 above)
Comment (BrandonOCasey) [inline src/js/player.js]: ok

File: src/js/player.js (hunk 8 above)
Comment (BrandonOCasey) [inline src/js/player.js]: using `ack` and `ripgrep` returns no results for `this.currentType_` now, so I don't think so.

File: src/js/player.js (hunk 7 above)
Comment (BrandonOCasey) [inline src/js/player.js]: added back in in the most recent commit

File: src/js/player.js [hunk 9]
```diff
@@ -2264,11 +2266,26 @@ class Player extends Component {
       }
 
       this.changingSrc_ = false;
-      this.cache_.src = src_.src;
+      this.cache_.source = src[0];
+      this.cache_.src = src_;
```
Comment (gkatsev) [inline src/js/player.js]: this should be `src_.src`.

**Comment by gkatsev** [comment]
Also, should we clear out the source caches if we error out? So, if everything worked but then you do `player.src([])`, `player.currentSources()` still returns the previously loaded sources.

**Comment by BrandonOCasey** [comment]
Maybe we should reset the source cache at the top of the function if it is not being used as a getter?

File: src/js/player.js (hunk 9 above)
Comment (BrandonOCasey) [inline src/js/player.js]: your right

**Comment by gkatsev** [comment]
Making the executive decision to merge even though travis is being silly. Hopefully master will be fine.
